src/controllers/keyboard/keyboardcontrollerpresetfilehandler.cpp

```diff
@@ -0,0 +1,140 @@
+#include "util/cmdlineargs.h"
+#include "util/compatibility.h"
+#include "control/controlobject.h"
+#include "controllers/keyboard/keyboardcontrollerpresetfilehandler.h"
+#include "controllers/keyboard/layoututils.h"
+
+
+KeyboardControllerPresetFileHandler::KeyboardControllerPresetFileHandler() {
+
+}
+
+KeyboardControllerPresetFileHandler::~KeyboardControllerPresetFileHandler() {
+
+}
+
+ControllerPresetPointer KeyboardControllerPresetFileHandler::load(const QDomElement root, const QString deviceName) {
+    if (root.isNull()) {
+        return ControllerPresetPointer();
+    }
+
+    QDomElement controller = getControllerNode(root, deviceName);
+    if (controller.isNull()) {
+        return ControllerPresetPointer();
+    }
+
+    KeyboardControllerPreset* preset = new KeyboardControllerPreset();
+
+    // Superclass handles parsing <info> tag and script files
+    parsePresetInfo(root, preset);
+    addScriptFilesToPreset(controller, preset);
+
+    // Iterate through each <group> block in the XML
+    QDomElement group = controller.firstChildElement("group");
+    while (!group.isNull()) {
+        QString groupName = group.attributeNode("name").value();
+
+        // Iterate through each <control> element inside the current <group> block
+        QDomElement control = group.firstChildElement("control");
+        while (!control.isNull()) {
+
+            // All keyseq elements (also the overloaded elements, that may
+            // not be used for the user's current keyboard layout)
+            QList<KbdControllerPresetKeyseq> keyseqsRaw;
+
+            // Create one Keyseq struct instance for each <keyseq> element inside the
+            // current <control> element and append it to keyseqsRaw
+            QDomElement keyseq_element = control.firstChildElement("keyseq");
+            while(!keyseq_element.isNull()) {
+                const QDomAttr finalAttr = keyseq_element.attributeNode("final");
+                bool isFinal = !finalAttr.isNull() && finalAttr.value() == "1";
+                keyseqsRaw.append(
+                        {
+                                keyseq_element.text(),                            // Key sequence
```

Comment:
Hmm, it looks like designated initializers didn't make it into C++11 and were only supported in C99. We could make a constructor for `KbdControllerPresetKeyseq` or just initialize it like this:

```c++
KbdControllerPresetKeyseq rawKeyseq;
rawKeyseq.keysequence = keyseq_element.text();
rawKeyseq.lang = keyseq_element.attributeNode("lang").value();
rawKeyseq.scancode = keyseq_element.attributeNode("scancode").value();

keyseqsRaw.append(data);
```

Do you know any other options?